core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -685,6 +692,7 @@ class ProducerStateManager(
 
     if (logEndOffset != mapEndOffset) {
       producers.clear()
+      _producerIdCount = 0
```

Comment:
Please move this together in a function. This coupling between the metric value `producers` is bug prone where someone may accidentally update one without updating the other. One way to avoid such bugs in future to always update the producer map using methods where count would be updated too (ideally only those functions should mutate the number of producers). 

Similar comment for adding to the `producers` as well.

Reply:
good point. I've coupled the two and made the change as you suggested.